tensorflow_federated/python/core/utils/execution_tracing.py

```diff
@@ -0,0 +1,271 @@
+
+      return sync_fn
+
+  def span(self, unused_scope, unused_sub_scope, **unused_trace_opts):
+    """"No-op implementation of span."""
+
+    class NoSuchError(Exception):
+      pass
+
+    # contextlib.suppress(NoSuchError) is used to create a no-op Context and
+    # should be replaced by contextlib.nullcontext when it becomes available.
+    return contextlib.suppress(NoSuchError)
+
+  def task_trace_context(self):
+    """No-op implementation of task_trace_context."""
+
+    class NoSuchError(Exception):
+      pass
+
+    # contextlib.suppress(NoSuchError) is used to create a no-op Context and
+    # should be replaced by contextlib.nullcontext when it is available.
+    return contextlib.suppress(NoSuchError)
+
+  def propagate_trace_context_task_factory(self, loop, coro):
+    """No-op implementation of propagate_trace_context_task_factory."""
+    return asyncio.tasks.Task(coro, loop=loop)
+
+  def run_coroutine_threadsafe_in_ambient_trace_context(self, coro, loop):
+    """No-op implementation of run_coroutine_threadsafe_in_ambient_trace_context."""
+    return asyncio.run_coroutine_threadsafe(coro, loop)
+
+  def run_coroutine_threadsafe_in_task_trace_context(self, coro, loop):
+    """No-op implementation of run_coroutine_threadsafe_in_ambient_trace_context."""
+    return asyncio.run_coroutine_threadsafe(coro, loop)
+
+  def run_coroutine_in_ambient_trace_context(self, coro):
+    """No-op implementation of run_coroutine_in_ambient_trace_context."""
+    return coro
+
+  @classmethod
+  def _is_traceable(cls, method_name, regex, override=False):
+    if override:
+      return True
+    return regex.search(method_name)
+
+  def _trace_method(self, fn, call_or_retr):
+    ctx = getattr(fn, '__self__', None)
+
+    def _log_fn(*args, **kwargs):
+      prefix_str = self._prefix(ctx, class_name)
+      main_str = "{m} {c}".format(m=method_name, c=call_or_retr)
+      suffix_args = " ".join(self.format_object(a) for a in args)
+      suffix_kwargs = " ".join(self.format_object(v) for k,v in kwargs.items())
+      logging.debug("{p} {m} {sa} {sk}".format(
+          p=prefix_str,
+          m=main_str,
+          sa=suffix_args,
+          sk=suffix_kwargs))
+
+    return _log_fn
+
+  def _prefix(self, ctx, type_):
+    indentation = '  ' * self._current_call_level
+    template = "{indentation}{type} {hash}"
+    if ctx is None:
+      return template.format(indentation=indentation, type=type_, hash="NO_ID")
+    return template.format(indentation=indentation, type=type_, hash=hash(ctx))
+
+  def _format_object(self, x):
+    strategy = self._formatting_strategies.get(type(x), self._default_strategy)
+    return strategy(x)
+
+  def _format_anon_tuple(self, tup):
+    names = tup._name_array
+    values = tup._element_array
+    fmt_strs = []
+    for n, v in zip(names, values):
+      fmt_strs.append("{}: {}".format(n, self.format_object(v)))
+    return "({})".format(", ".join(fmt_strs))
+
+  #
+  # Registration of built-in types
```

Comment:
perhaps reformat comment to match elsewhere

Reply:
this was matching the original RFC file, but I agree